File: src/1143_Longest_Common_Subsequence/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    // 经典的动态规划题目: 要建立二维DP数组dp[i][j]
    // 关键是dp中i和j的意义, i和j分别代表text1中0~i这段substr1和text2中0~j这段substr2
    // dp[i][j]代表的是substr1和substr2的最长common string
    // 这里的subsequence其实提示了可以用for loop来解决, 
    // 因为for loop是依次遍历index的, 这就保证了次序
    // Runtime: 36 ms, faster than 26.67%
    // Memory Usage: 13.2 MB, less than 5.54%
    int longestCommonSubsequence(string text1, string text2) {
        int rtn, len1 = text1.size(), len2 = text2.size();
        // 然后dp数组的更新条件是:
        // 如果text1[i]与text2[j]相等, 那么dp[i][j]就等于dp[i-1][j-1] + 1
        // 如果text1[i]与text2[j]不等, 那么dp[i][j]就等于dp[i-1][j]或者dp[i][j-1]
        // 中的较大者, 这等价于用稍微短一个长度的text1’或者text2‘与另一个text进行比较

        // 有个小trick, 由于dp的更新条件出现了i-1和j-1, 所以为了避免逻辑判断让代码简洁
        // 初始化dp时为text1和text2都多申请一个, 就可以巧妙处理边界情况, 只需注意index
        // vector的默认值等于0, 所以正好
        vector<vector<int>> dp(len1+1, vector<int>(len2+1));

        for (int i = 1; i <= len1; ++i) {
            for (int j = 1; j <= len2; ++j) {
                // 稍微注意一下i和j到text1和text2的index映射
                if (text1[i-1] == text2[j-1])
                    dp[i][j] = dp[i-1][j-1] + 1;
                else
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
            }
        }
        return dp[len1][len2];
    }
};

int main(int argc, const char * argv[]) {
    // insert code here...
    Solution gua;
    string text1, text2;
    int rtn;
    rtn = gua.longestCommonSubsequence(text1 = "abcde", text2 = "ace");
    cout << "expect value = 3, actual value = " << rtn << endl;

    rtn = gua.longestCommonSubsequence(text1 = "abc", text2 = "def");
    cout << "expect value = 0, actual value = " << rtn << endl;
    return 0;
}

File: src/1143_Longest_Common_Subsequence/main.py

class Solution:
    # Runtime: 376 ms, faster than 84.02% 
    # Memory Usage: 14.3 MB, less than 94.92%
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # 来压缩一下空间
        size1, size2 = len(text1), len(text2)
        dp = [0 for j in range(size2+1)]

        for i in range(1, size1+1):
            prev = dp[0] # prev保存的是dp[i-1][j-1]的值，压缩之后变成dp[j-1]
            for j in range(1, size2+1):
                #     a b c d e
                #   0 0 0 0 0 0
                # a 0 0 0 0 0 0
                # c 0 0 0 0 0 0
                # e 0 0 0 0 0 0
                
                # 注意，当保存信息到dp[i][j]时，实际上覆盖了dp[i-1][j]的值，让右侧的计算不正确，
                # 所以要先保留dp[i-1][j]的值，而最早的dp[i-1][j]是0列的值，始终为0
                if text1[i-1] == text2[j-1]:
                    newV = prev + 1 # i被压缩掉了
                else:
                    newV = max(dp[j-1], dp[j]) # 这两部分的值还是不变
                prev = dp[j] # 在即将更新dp[i][j]时，保留当前位置dp[i-1][j]的值，作为下一轮的对角线值
                dp[j] = newV # 此时再更新dp[i][j]的值，dp[i-1][j]已经保留，一切OK
                # if text1[i-1] == text2[j-1]:
                #     dp[j] = dp[j-1] + 1 # i被压缩掉了
                # else:
                #     dp[j] = max(dp[j-1], dp[j])
        return dp[size2]
    
    # Runtime: 428 ms, faster than 56.26%
    # Memory Usage: 23 MB, less than 31.97% 
    def longestCommonSubsequence1(self, text1: str, text2: str) -> int:
        # 两个字符串的操作，dp[i][j]常用于表示text1中i长度的substr和text2中j长度的substr的LCS
        # 先用二维数组，一会儿再压缩空间
        size1, size2 = len(text1), len(text2)
        
        # 行列分别是text1和text2的substr长度，第0行和第0列初始化为0
        dp = [[0 for j in range(size2+1)] for i in range(size1+1)]
        
        # 更新时从左到右，从上到下遍历
        for i in range(1, size1+1):
            for j in range(1, size2+1):
                if text1[i-1] == text2[j-1]:
                    # 出现相同字符的话，直接在上一轮的dp数组基础上+1
                    dp[i][j] = dp[i-1][j-1] + 1
                else:
                    # 不相同的话就看是去掉text1还是text2
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
        
        return dp[size1][size2]

if __name__ == "__main__":
  gua = Solution()
  rtn = gua.longestCommonSubsequence(text1 = "abcba", text2 = "abcbcba")
  print("expect result = 5, actual result = %d " % rtn )

  rtn = gua.longestCommonSubsequence(text1 = "abc", text2 = "abc")
  print("expect result = 3, actual result = %d " % rtn )

  rtn = gua.longestCommonSubsequence(text1 = "abcde", text2 = "ace")
  print("expect result = 3, actual result = %d " % rtn )

  rtn = gua.longestCommonSubsequence(text1 = "abc", text2 = "def")
  print("expect result = 0, actual result = %d " % rtn )

